### PR TITLE
Complete refactor to make it work

### PR DESCRIPTION
**Pull request by gideonthomas**

### CODE REVIEW

**Comment by gideonthomas** [comment]
@humphd can you take a look and see if everything looks ok?

**Comment by humphd** [comment]
4-space indents, please.

File: parser.js [hunk 1]
```diff
@@ -1,273 +1,51 @@
...
-            }
-            //character relative to start of the line
-            msg[1] = msg[1] - lineBeginStart;
-            msg[2] = msg[2] - lineBeginEnd;
-            
-            //line number for end of error
-            msg[4] = lineCount;
+    function parse(input) {
+        var parse = require("slowparse/slowparse");
+        var errorMessages = require("text!slowparse/locale/en_US.json");
```
Comment (humphd) [inline parser.js]: Let's move these requires outside the function call, up near the top of this module.

File: errorDisplay.js [hunk 2]
```diff
@@ -3,108 +3,87 @@
...
-
-		if(!allMarks.length) {
-			editor._codeMirror.markText({line: lineStart, ch: charStart},
-				{line: lineEnd, ch: charEnd},
-				{className: "errorHighlight"});
-		}
+
+	function initGutter() {
+		var errorGutter = document.createElement("div");
+		errorGutter.className = "errors";
```
Comment (humphd) [inline errorDisplay.js]: I'm confused about `errorGutter`.  You create the element here, set a `className`, then never use it again, nor add it to the DOM.  Why?

File: errorDisplay.js [hunk 3]
```diff
@@ -3,108 +3,87 @@
...
-	exports.showGutter = showGutter;
-	exports.removeGutter = removeGutter;
-	exports.removeWidget = removeWidget;
-});
\ No newline at end of file
+
+	function hideDescription() {
+		if(currentErrorWidget) {
+			currentErrorWidget.clear();
+			currentErrorWidget = undefined;
```
Comment (humphd) [inline errorDisplay.js]: Use `null` for your rvalue unless you actually want `undefined` (which you don't).  `undefined` means that it's never been defined, not that you've defined it as `undefined`.

File: main.js [hunk 4]
```diff
@@ -4,64 +4,64 @@
...
-                results.push(result);
-            }else{
-                clearAllErrors();
-            }
-            BottomDisplayVar.update(text);
+        var editor = EditorManager.getActiveEditor();
+        var error;
+        var html;
+
+        if(!editor || editor.document.getLanguage()._name !== "HTML") {
```
Comment (humphd) [inline main.js]: It's concerning that you're having to use a private member here, `_name`.  @K-LV was doing something like this using a public `id` field, I wonder if you could too?

File: parser.js [hunk 5]
```diff
@@ -1,273 +1,51 @@
...
+                .replace(/\[\[closeTag\.name\]\]/, replaceText.closeTag)
+                .replace(/\[\[openTag\.name\]\]/, replaceText.openTag)
+                .replace(/\[\[attribute\.name\.value\]\]/, replaceText.attribute)
+                .replace(/\[\[cssValue\.value\]\]/, replaceText.cssValue)
+                .replace(/\[\[cssSelector\.selector\]\]/, replaceText.cssSelector)
+                .replace(/\[\[cssProperty\.property\]\]/, replaceText.cssProperty)
+                .replace(/\[\[cssKeyword\.value\]\]/, replaceText.cssKeyword)
+                .replace(/\[\[name\]\]/, replaceText.name)
+                .replace(/\[\[error\.msg\]\]/, replaceText.error)
+                .replace(/\[\[value\]\]/, replaceText.value);
```
Comment (humphd) [inline parser.js]: Could you not use Mustache or some other template rendering to do this, vs. having to regex all this?  Isn't that the point of why they did it in that style?

**Comment by humphd** [comment]
r- for now.  See my comments.  Thanks for working on this.

**Comment by gideonthomas** [comment]
@humphd does this look good now?

File: errorDisplay.js [hunk 6]
```diff
@@ -2,109 +2,85 @@
...
+    function initGutter() {
+        EditorManager.getActiveEditor()._codeMirror.setOption("gutters", ["errors"]);
+    }
+
+    function highlight(line) {
+        if(!line) {
+           return;
+        }
+
+        var editor = EditorManager.getActiveEditor();
```
Comment (humphd) [inline errorDisplay.js]: You have a lot of code that does this, can you put it into a function?  `getActiveEditor()` should return it, and use everywhere.

Comment (humphd) [inline errorDisplay.js]: Also add a function for getting the CodeMirror instance, since you do that a lot too.

File: errorDisplay.js [hunk 7]
```diff
@@ -2,109 +2,85 @@
...
+        var description = document.createElement("div");
+        description.className = "errorPanel";
+        description.innerHTML = Mustache.render(lineWidgetHTML, {"error": error.message});
+        var options = {coverGutter: false, noHScroll: false, above: false, showIfHidden: false};
+
+        currentErrorWidget = editor._codeMirror.addLineWidget(error.line, description, options);
+    }
+
+    function hideDescription() {
+        if(currentErrorWidget) {
```
Comment (humphd) [inline errorDisplay.js]: Invert this, you should return if `!currentErrorWidget` and unindent the rest of the code.

**Comment by humphd** [comment]
A couple nits, looks good otherwise.  I've not tested it functionally.  r=me with this stuff fixed.

**Comment by gideonthomas** [comment]
done :)

**Comment by gideonthomas** [comment]
@KeniMorri can you merge in?

**Pull request merged**